op-node/node/safedb/safedb.go

```diff
@@ -0,0 +1,192 @@
+package safedb
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+	"slices"
+	"sync"
+
+	"github.com/cockroachdb/pebble"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+var (
+	ErrNotFound     = errors.New("not found")
+	ErrInvalidEntry = errors.New("invalid db entry")
+)
+
+const (
+	// Keys are prefixed with a constant byte to allow us to differentiate different "columns" within the data
+	keyPrefixSafeByL1BlockNum byte = 0
+)
+
+var (
+	SafeByL1BlockNumKey = uint64Key{prefix: keyPrefixSafeByL1BlockNum}
+)
+
+type uint64Key struct {
+	prefix byte
+}
+
+func (c uint64Key) Of(num uint64) []byte {
+	key := make([]byte, 0, 9)
+	key = append(key, c.prefix)
+	key = binary.BigEndian.AppendUint64(key, num)
+	return key
+}
+func (c uint64Key) Max() []byte {
+	return c.Of(math.MaxUint64)
+}
+
+func (c uint64Key) IterRange() *pebble.IterOptions {
+	return &pebble.IterOptions{
+		LowerBound: c.Of(0),
+		UpperBound: c.Max(),
+	}
+}
+
+type SafeDB struct {
+	// m ensures all read iterators are closed before closing the database by preventing concurrent read and write
+	// operations (with close considered a write operation).
+	m   sync.RWMutex
+	log log.Logger
+	db  *pebble.DB
+
+	writeOpts *pebble.WriteOptions
+
+	closed bool
+}
+
+func SafeByL1BlockNumValue(l1 eth.BlockID, l2 eth.BlockID) []byte {
+	val := make([]byte, 0, 72)
+	val = append(val, l1.Hash.Bytes()...)
+	val = append(val, l2.Hash.Bytes()...)
+	val = binary.BigEndian.AppendUint64(val, l2.Number)
+	return val
+}
+
+func DecodeSafeByL1BlockNum(key []byte, val []byte) (l1 eth.BlockID, l2 eth.BlockID, err error) {
+	if len(key) != 9 || len(val) != 72 {
+		err = ErrInvalidEntry
+		return
+	}
+	copy(l1.Hash[:], val[:32])
+	l1.Number = binary.BigEndian.Uint64(key[1:])
+	copy(l2.Hash[:], val[32:64])
+	l2.Number = binary.BigEndian.Uint64(val[64:])
+	return
```

Comment:
Should the value of `key[0]` be asserted here, since it's part of the input?

Reply:
Yeah seems like a good idea. Added.